extmod/mbedtls/mbedtls_alt.c

```diff
@@ -0,0 +1,84 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright The Mbed TLS Contributors
+ * Copyright (c) 2024 Damien P. George
+ *
+ * This file provides default fallback functions for use with alternate
+ * cryptography functions implemented in Python.
+ */
+#if MICROPY_PY_SSL_ECDSA_SIGN_ALT
+#include <string.h>
+#define MBEDTLS_ALLOW_PRIVATE_ACCESS
+#include "mbedtls/platform.h"
+#include "mbedtls/ssl.h"
+#include "mbedtls/error.h"
+#include "mbedtls/ecdsa.h"
+#include "mbedtls/asn1write.h"
+
+extern int mbedtls_ecdsa_sign_alt(mbedtls_ecp_group *grp, mbedtls_mpi *r, mbedtls_mpi *s,
+    const mbedtls_mpi *d, const unsigned char *buf, size_t blen,
+    int (*f_rng)(void *, unsigned char *, size_t), void *p_rng);
+
+/*
+ * Compute and write signature
+ */
+int __wrap_mbedtls_ecdsa_write_signature(mbedtls_ecdsa_context *ctx,
+    mbedtls_md_type_t md_alg,
+    const unsigned char *hash, size_t hlen,
+    unsigned char *sig, size_t sig_size, size_t *slen,
+    int (*f_rng)(void *, unsigned char *, size_t),
+    void *p_rng) {
+
+    (void)md_alg;
+
+    mbedtls_mpi r, s;
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    unsigned char buf[MBEDTLS_ECDSA_MAX_LEN] = { 0 };
+    unsigned char *p = buf + sizeof(buf);
+
+    if (f_rng == NULL) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    }
+
+    // Check if curve is supported for ECDSA.
+    if (!mbedtls_ecdsa_can_do(ctx->grp.id) || ctx->grp.N.p == NULL) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    }
+
+    mbedtls_mpi_init(&r);
+    mbedtls_mpi_init(&s);
+
+    // Try signing with the alternate function first, then fallback to the default.
+    if (mbedtls_ecdsa_sign_alt(&ctx->grp, &r, &s, &ctx->d, hash, hlen, f_rng, p_rng) != 0) {
+        MBEDTLS_MPI_CHK(mbedtls_ecdsa_sign(&ctx->grp, &r, &s, &ctx->d, hash, hlen, f_rng, p_rng));
```

Comment:
Note to avoid duplicating a lot of code `MBEDTLS_ECDSA_SIGN_ALT` is NOT defined to keep the default `mbedtls_ecdsa_sign` to use as a fallback function. However, `mbedtls_ecdsa_sign` can't be wrapped, because it's called internally to its object file, so we wrap `mbedtls_ecdsa_write_signature` and `mbedtls_ecdsa_read_signature` instead.

Reply:
It's annoying there's no way to keep the default implementation without a linker wrap. I just went down the rabbit hole of looking at replacing the `mbedtls_eckey_info` structure instead (as it's the only place inside mbedTLS which calls through to here), but I can't see a different way there either.

Could you please put some of your rationale from the comment here into the top-level comment for `__wrap_mbedtls_ecdsa_write_signature`? That way it'll be clearer for whoever touches this next.

Reply:
@projectgus I also tried many different things, this is the only way right now. And yes it's extremely annoying, I don't know why they didn't just make the functions weak like everyone else and saved us the trouble.

Reply:
Done.